common/src/gas_provider/mod.rs

```diff
@@ -207,6 +207,15 @@ pub trait Tree {
     /// for example, for `GasNode::ReservedLocal`.
     fn get_lock(key: Self::Key) -> Result<Self::Balance, Self::Error>;
 
+    /// Reserve some value from underlying balance.
+    ///
+    /// Used in gas reservation feature.
+    fn reserve(
```

Comment:
To avoid `Into::into` everywhere (and also avoid `impl Into<Self::Key>` in fns) I'd add one more associated type to GasHandler trait with ReservationId, and use in storage some generic enum for `GasNodeType<NodeId, ReservationId>`